lib/web/apiserver_test.go

```diff
@@ -2002,6 +2003,106 @@ func TestGetMFADevices(t *testing.T) {
+
+	endpoint := pack.clt.Endpoint("webapi", "mfa", "authnchallenge")
+	re, err := pack.clt.PostJSON(context.Background(), endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var chal auth.MFAAuthenticateChallenge
+	err = json.Unmarshal(re.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
+}
+
+func TestCreateAuthenticateChallenge_WithToken(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	ap, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorOTP,
+	})
+	require.NoError(t, err)
+	err = env.server.Auth().SetAuthPreference(ctx, ap)
+	require.NoError(t, err)
+
+	// Create a user with a TOTP device.
+	username := "llama"
+	proxy.createUser(ctx, t, username, "root", "password", "some-otp-secret")
+
+	// Acquire a start token.
+	startToken, err := types.NewUserToken("some-token-id")
+	require.NoError(t, err)
+	startToken.SetUser(username)
+	startToken.SetSubKind(auth.UserTokenTypeRecoveryStart)
+	startToken.SetExpiry(env.clock.Now().Add(5 * time.Minute))
+	_, err = env.server.Auth().Identity.CreateUserToken(ctx, startToken)
+	require.NoError(t, err)
+
+	// Call the endpoint.
+	clt := proxy.newClient(t)
+	endpoint := clt.Endpoint("webapi", "mfa", "token", startToken.GetName(), "authnchallenge")
+	res, err := clt.PostJSON(ctx, endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var chal auth.MFAAuthenticateChallenge
+	err = json.Unmarshal(res.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
+}
+
+func TestCreateAuthenticateChallenge_WithUserCreds(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	// Create a user with a TOTP device, with second factor preference to OTP only.
+	pack := proxy.authPack(t, "llama@example.com")
+
+	requestBody := client.MFAChallengeRequest{
+		User: pack.user,
+		Pass: pack.password,
+	}
+
+	// Test private endpoint.
+	endpoint := pack.clt.Endpoint("webapi", "u2f", "password", "changerequest")
+	re, err := pack.clt.PostJSON(ctx, endpoint, requestBody)
+	require.NoError(t, err)
+
+	var chal auth.MFAAuthenticateChallenge
+	err = json.Unmarshal(re.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
+
+	// Test public endpoint.
+	clt := proxy.newClient(t)
+	endpoint = clt.Endpoint("webapi", "mfa", "login", "begin")
+	res, err := clt.PostJSON(ctx, endpoint, requestBody)
+	require.NoError(t, err)
+
+	err = json.Unmarshal(res.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
```

Comment:
Loop over the various endpoints?

```suggestion
tests := []struct{
	name string
	clt client
	ep  []string
}{
	{
		name: "/webapi/u2f/password",
		clt: pack.clt,
		ep: []string{"webapi", "u2f", "password"}
	},
	// Entry for /webapi/mfa/authnchallenge
	// Entry for /webapi/mfa/login/begin
}
for _, test := range tests {
	endpoint = test.clt.Endpoint(test.ep)
	res, err := test.clt.PostJSON(ctx, endpoint, requestBody)
	require.NoError(t, err)

	err = json.Unmarshal(res.Bytes(), &chal)
	require.NoError(t, err)
	require.True(t, chal.TOTPChallenge)
	require.Empty(t, chal.U2FChallenges)
	require.Empty(t, chal.WebauthnChallenge)
}
```